### DISCUSSION TITLE
Graceful shutdown FastAPI integration

### DISCUSSION THREAD

**Discussion by dimbler** [ag2ai/faststream#979] [General]
Dear Team,
Thank you for great library.
Please clarify, how can support on_shutdown event in FastAPI integration?
This question related to https://github.com/airtai/faststream/issues/965

**Comment by Lancetnik** [comment]
We are working on this feature. For now, you can add some middleware, taking asyncio.Lock while message is processed and wait for release it in your shutdown event.

**Reply by dimbler** [reply]
Thank you for rapid answer.
Probably it's only valuable part of shutdown process, but also has another part of shutdown process (usually in fastapi shutdown event) when perform all waiting wasks (maybe scheduled)
So we are waiting for the desired feature

**Comment by Lancetnik** [comment]
@dimbler sorry, I just forgot about this discussion
Did you saw the [0.3.5 release](https://github.com/airtai/faststream/releases/tag/0.3.5)? It added the thing you want

**Reply by dimbler** [reply]
Dear @Lancetnik,
Thank you for assistance, last time probably not accurate describe my idea, but originally desire catch a shutdown event to have a time for end all running action like fastapi `@app.on_event("shutdown")`

**Reply by Lancetnik** [reply]
Oh, sorry, didn't got it. `graceful_timeout` options guarantee that all already running handlers should be completed before shutdown. But, if you want to make smth custom before shutdown, unfortunately you have no option for now.
But, I can add `on_shutdown` hook in router the same with `after_startup` way. Will it close your problem?

**Reply by dimbler** [reply]
Thank you very much, on_shutdown mechanism looks brilliant idea

**Reply by Lancetnik** [reply]
Did you see #1184? It is in main branch already

**Reply by dimbler** [reply]
Probably din't understand how can use it? after_startup decorator get a function to call it, but on_shutdown it a fastapi event, how i can use it in application?

**Reply by Lancetnik** [reply]
> Probably din't understand how can use it? after_startup decorator get a function to call it, but on_shutdown it a fastapi event, how i can use it in application?

This reason it calls like 'on_broker_shutdown'

**Reply by dimbler** [reply]
Sorry, don't get it. Do you mean:
```
@router.on_broker_shutdown
async def shutdown_process(app: FastAPI):
    logger.info(f"Shutdown app on {Path().resolve().joinpath('app')}")
```
